spinalcordtoolbox/reports/qc.py

```diff
@@ -337,17 +349,17 @@ def _make_QC_image_for_3d_volumes(self, img, mask, slice_orientation):
         logger.info(self.qc_report.qc_params.abs_bkg_img_path())
         self._save(fig, self.qc_report.qc_params.abs_bkg_img_path(), dpi=self.qc_report.qc_params.dpi)
 
-        for action in self.action_list:
+        fig = Figure()
+        fig.set_size_inches(size_fig[0], size_fig[1], forward=True)
+        FigureCanvas(fig)
+        for i, action in enumerate(self.action_list):
             logger.debug('Action List %s', action.__name__)
             if self._stretch_contrast and action.__name__ in ("no_seg_seg",):
-                print("Mask type %s" % mask.dtype)
-                mask = self._func_stretch_contrast(mask)
-            fig = Figure()
-            fig.set_size_inches(size_fig[0], size_fig[1], forward=True)
-            FigureCanvas(fig)
-            ax = fig.add_axes((0, 0, 1, 1))
-            action(self, mask, ax)
-            self._save(fig, self.qc_report.qc_params.abs_overlay_img_path(), dpi=self.qc_report.qc_params.dpi)
+                print("Mask type %s" % mask[i].dtype)
+                mask[i] = self._func_stretch_contrast(mask[i])
+            ax = fig.add_axes((0, 0, 1, 1), label=str(i))
+            action(self, mask[i], ax)
+        self._save(fig, self.qc_report.qc_params.abs_overlay_img_path(), dpi=self.qc_report.qc_params.dpi)
```

Comment:
I tried taking a look at this section using the commands from my previous review, and I noticed that `self.action_list` had 3 entries, but `mask` had 4 volumes, so I think the last volume never gets used.

![Screenshot from 2021-07-22 11-55-08](https://user-images.githubusercontent.com/16181459/126669931-15b47e36-f5c7-402a-a715-ffa7c1df3ca7.png)

The 4 volumes correspond to `fname_list[1:4]`:

![Screenshot from 2021-07-22 12-01-41](https://user-images.githubusercontent.com/16181459/126670947-2bf3d919-712a-4831-9397-fc6b9e073d4b.png)

So, the 4th volume is `t2_seg_centerline_extrapolated_smooth.nii.gz`... but it's repeated? So how did this get here... (Answer: Link to other review comment.)

Reply:
Yeah it is repeated, the problem is that the last volume is used to select the sagittal slice for the Qc, so I want it to be the smoothed centerline, but if I leave it at the end, it will be displayed over PMJ and the mask (not the desired display), this was the solution that I found for this, there may be a better way if you have any ideas 🙂

Reply:
Ohh, I see. Thank you for clarifying! I'll try to take a look and see if there's a better way.

If there isn't a better way, perhaps at the very least we could leave a comment explaining this behavior. (Tracing the QC module is quite a pain, isn't it... :sweat_smile:. The wrapped functions are a little hard to understand! So, if there are any comments we can leave here that might help people navigate this code in the future, that might be good.)

Reply:
Found the comment, I knew I specified it somewhere 😅:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/754cbe86eec066035481a328684f768d5f0f9875/spinalcordtoolbox/scripts/sct_process_segmentation.py#L408

I could also add it in qc.py

Reply:
You were absolutely right... this is necessary, sadly. :sweat_smile: 

I've added some suggestions for comments to make this a little clearer perhaps: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3465#discussion_r675561784 and https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3465#discussion_r675574281)